pkg/controller/v1beta1/inferenceservice/reconcilers/deployment/deployment_reconciler.go

```diff
@@ -398,33 +398,36 @@ func (r *DeploymentReconciler) Reconcile(ctx context.Context) ([]*appsv1.Deploym
 		var opErr error
 		switch checkResult {
 		case constants.CheckResultCreate:
-			opErr = r.client.Create(ctx, deployment)
+			opErr = r.client.Create(ctx, desiredDep)
 		case constants.CheckResultUpdate:
-			curJson, err := json.Marshal(deployment)
-			if err != nil {
-				return nil, err
-			}
+			curDeployment := existingDep.DeepCopy()
+			modDeployment := desiredDep.DeepCopy()
 
 			// To avoid the conflict between HPA and Deployment,
 			// we need to remove the Replicas field from the deployment spec
 			// For none autoscaler, it should not remove replicas
-			modDeployment := deployment.DeepCopy()
 			if modDeployment.Annotations[constants.AutoscalerClass] != string(constants.AutoscalerClassNone) {
 				modDeployment.Spec.Replicas = nil
+				curDeployment.Spec.Replicas = nil
+			}
+
+			curJson, err := json.Marshal(curDeployment)
+			if err != nil {
+				return nil, err
 			}
 
 			modJson, err := json.Marshal(modDeployment)
 			if err != nil {
 				return nil, err
 			}
 			// Generate the strategic merge patch between the current and modified JSON
-			patchByte, err := strategicpatch.StrategicMergePatch(curJson, modJson, appsv1.Deployment{})
+			patchByte, err := strategicpatch.CreateTwoWayMergePatch(curJson, modJson, appsv1.Deployment{})
```

Comment:
see https://kubernetes.io/docs/tasks/manage-kubernetes-objects/update-api-object-kubectl-patch/#use-a-json-merge-patch-to-update-a-deployment. 

How `strategicpatch.CreateTwoWayMergePatch` is different from `client.MergeFrom()` which returns the merge patch object ?

Reply:
Can you help test the case when fields are updated/removed, for example replacing tcp readiness probe kserve added by default to http readiness probe.

Reply:
Cc @spolti to review

Reply:
@andyi2it, does it work for removing envs or updating secrets?
Please take a look at this PR:  https://github.com/kserve/kserve/pull/4297

I guess this would fix the envs removal as well. 
The `CreateTwoWayMergePatch` treats lists as generic JSON objects.

Could you copy the unit test I added in the other PR here? That tests the addition and removal of envs.